### PR TITLE
React to HttpAbstractions#146 changes

### PR DESCRIPTION
**Pull request by ajaybhargavb**
@pranavkm Please review.

### CODE REVIEW

**Comment by pranavkm** [comment]
Discussed with @ajaybhargavb. We can reduce the size of the PR since this change is also attempting to address support for multi part mime parsing.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/FormCollectionValueProvider.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        MetadataAwareValueProvider<IFormDataValueProviderMetadata>,
+        IEnumerableValueProvider
+    {
+        private readonly CultureInfo _culture;
+        private PrefixContainer _prefixContainer;
+        private readonly Func<Task<IFormCollection>> _valuesFactory;
+        private IFormCollection _values;
+
+        /// <summary>
+        /// Creates a NameValuePairsProvider wrapping an existing set of key value pairs.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/FormCollectionValueProvider.cs]: Doc comment is wrong

File: src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/FormCollectionValueProvider.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            return prefixContainer.ContainsPrefix(prefix);
+        }
+
+        public virtual async Task<IDictionary<string, string>> GetKeysFromPrefixAsync([NotNull] string prefix)
+        {
+            var prefixContainer = await GetPrefixContainerAsync();
+            return prefixContainer.GetKeysFromPrefix(prefix);
+        }
+
+        public override async Task<ValueProviderResult> GetValueAsync([NotNull] string key)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/FormCollectionValueProvider.cs]: Missing doc comments here

File: src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/FormValueProviderFactory.cs [hunk 3]
```diff
@@ -19,8 +19,8 @@ public IValueProvider GetValueProvider([NotNull] ValueProviderFactoryContext con
             if (IsSupportedContentType(request))
             {
                 var culture = GetCultureInfo(request);
-                return new ReadableStringCollectionValueProvider<IFormDataValueProviderMetadata>(
-                    () => request.GetFormAsync(),
+                return new FormCollectionValueProvider(
+                    () => request.ReadFormAsync(),
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/FormValueProviderFactory.cs]: extra lambda isn't needed here, you should be able to pass `request.ReadFormAsync`

File: src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/QueryStringValueProvider.cs [hunk 4]
```diff
@@ -11,13 +11,13 @@
 
 namespace Microsoft.AspNet.Mvc.ModelBinding
 {
-    public class ReadableStringCollectionValueProvider<TBinderMetadata> :
-        MetadataAwareValueProvider<TBinderMetadata>, IEnumerableValueProvider
-        where TBinderMetadata : IValueProviderMetadata
+    public class QueryStringValueProvider :
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/QueryStringValueProvider.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/QueryStringValueProvider.cs [hunk 5]
```diff
@@ -103,4 +103,4 @@ public CultureInfo Culture
             return _prefixContainer;
         }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/QueryStringValueProvider.cs]: doc comments?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/ValueProviders/QueryStringValueProviderTest.cs [hunk 6]
```diff
@@ -219,17 +219,17 @@ public static IEnumerable<object[]> RegisteredAsMetadataClasses
...
             var result = valueProvider.Filter(metadata);
 
             // Assert
             Assert.NotNull(result);
-            Assert.IsType<ReadableStringCollectionValueProvider<TestValueProviderMetadata>>(result);
+            Assert.IsType<QueryStringValueProvider>(result);
         }
 
-        private class TestValueProviderMetadata : IValueProviderMetadata
+        private class TestValueProviderMetadata : IQueryValueProviderMetadata
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/ValueProviders/QueryStringValueProviderTest.cs]: Is this still used anywhere?

Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/ValueProviders/QueryStringValueProviderTest.cs]: It is used in this test `FilterReturnsItself_ForAnyClassRegisteredAsGenericParam`.

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/ValueProviders/QueryStringValueProviderTest.cs]: cool beans

**Pull request closed without merging**